src/data/equiv/algebra.lean

```diff
@@ -190,6 +199,88 @@ protected def discrete_field [discrete_field β] : discrete_field α :=
 end instances
 end equiv
 
+structure monoid_equiv (M N : Type*) [monoid M] [monoid N] extends M ≃ N :=
+(hom : is_monoid_hom to_fun)
+
+infix ` ≃m `:50 := monoid_equiv
+
+namespace monoid_equiv
+
+variables [monoid α] [monoid β] [monoid γ]
+
+instance : has_coe_to_fun (α ≃m β) :=
+{ F := λ _, α → β,
+  coe := λ h, h.to_fun }
+
+instance is_monoid_hom (h : α ≃m β) :
+is_monoid_hom h := h.hom
+
+def refl (α : Type) [monoid α] : α ≃m α :=
+{ hom := is_monoid_hom.id,
+..equiv.refl _}
+
+def symm (h : α ≃m β) : β ≃m α :=
+{ hom := { map_one := by rw ←h.hom.map_one; exact (h.left_inv 1),
+  map_mul := λ n₁ n₂, function.injective_of_left_inverse h.left_inv begin
+   rw h.hom.map_mul, unfold equiv.symm, rw [h.right_inv, h.right_inv, h.right_inv], end }
+ ..h.to_equiv.symm}
+
+def trans (h1 : α ≃m β) (h2 : β ≃m γ) : (α ≃m γ) := {
+  hom := is_monoid_hom.comp _ _,
+  ..equiv.trans h1.to_equiv h2.to_equiv }
+
+instance symm.is_monoid_hom (h : α ≃m β) :
+is_monoid_hom h.symm := h.symm.hom
+
+end monoid_equiv
+
+structure group_equiv (α β : Type*) [group α] [group β] extends α ≃ β :=
+{ hom : is_group_hom to_fun}
+
+infix ` ≃g `:50 := group_equiv
```

Comment:
Let's make `g` subscript. 

Also all the monoid hom comments apply to the group hom lemmas

Reply:
Now switched to mul_equiv after conversation with Chris.